codegen/src/api/errors.rs

```diff
@@ -0,0 +1,171 @@
+    };
+
+    let match_body_items = errors.into_iter().map(|err| {
+        let docs = err.description();
+        let pallet_index = err.pallet_index;
+        let error_index = err.error_index;
+        let pallet_name = err.pallet;
+        let error_name = err.error;
+
+        quote! {
+            (#pallet_index, #error_index) => Some(ErrorDetails {
+                pallet: #pallet_name,
+                error: #error_name,
+                docs: #docs
+            })
+        }
+    });
+
+    ErrorDetails {
+        // A type we'll be returning that needs defining at the top level:
+        type_def: quote! {
+            pub struct ErrorDetails {
+                pub pallet: &'static str,
+                pub error: &'static str,
+                pub docs: &'static str,
+            }
+        },
+        // A function which will live in an impl block for our DispatchError,
+        // to statically return details for known error types:
+        dispatch_error_impl_fn: quote! {
+            pub fn details(&self) -> Option<ErrorDetails> {
+                if let Self::Module { index, error } = self {
+                    match (index, error) {
+                        #( #match_body_items ),*,
+                        _ => None
+                    }
+                } else {
+                    None
+                }
+            }
+        },
+    }
+}
+
+fn pallet_errors(
+    metadata: &RuntimeMetadataV14,
+) -> Result<Vec<ErrorMetadata>, InvalidMetadataError> {
+    let get_type_def_variant = |type_id: u32| {
+        let ty = metadata
+            .types
+            .resolve(type_id)
+            .ok_or(InvalidMetadataError::MissingType(type_id))?;
+        if let scale_info::TypeDef::Variant(var) = ty.type_def() {
+            Ok(var)
+        } else {
+            Err(InvalidMetadataError::TypeDefNotVariant(type_id))
+        }
+    };
+
+    let pallet_errors = metadata
+        .pallets
+        .iter()
+        .filter_map(|pallet| {
+            pallet.error.as_ref().map(|error| {
+                let type_def_variant = get_type_def_variant(error.ty.id())?;
+                Ok((pallet, type_def_variant))
+            })
+        })
+        .collect::<Result<Vec<(_, _)>, _>>()?;
+
+    let errors = pallet_errors
+        .iter()
+        .flat_map(|(pallet, type_def_variant)| {
+            type_def_variant.variants().iter().map(move |var| {
+                ErrorMetadata {
+                    pallet_index: pallet.index,
+                    error_index: var.index(),
+                    pallet: pallet.name.clone(),
+                    error: var.name().clone(),
+                    variant: var.clone(),
+                }
+            })
+        })
+        .collect();
+
+    Ok(errors)
+}
+
+#[derive(Clone, Debug)]
+pub struct ErrorMetadata {
+    pub pallet_index: u8,
+    pub error_index: u8,
+    pub pallet: String,
+    pub error: String,
+    variant: scale_info::Variant<scale_info::form::PortableForm>,
```

Comment:
dq: what is the reason for storing the variant for?

Reply:
ah, for the `description()`?

Reply:
Yup, that's it! I'm not sure why I did/left this now; it's just the doc string, which simplifies things a bit!